File: mobile/lib/shared/models/month_key.dart
/// Business date keys: transactionDateKey = YYYY-MM-DD, monthKey = YYYY-MM.
/// Always derive from the business date the user picked (Asia/Karachi),
/// never from device UTC instant, to keep monthly reports stable.
class MonthKey {
  static String fromDate(DateTime d) {
    if (d.day >= 11) {
      final nextMonth = DateTime(d.year, d.month + 1, 1);
      return '${nextMonth.year.toString().padLeft(4, '0')}-${nextMonth.month.toString().padLeft(2, '0')}';
    } else {
      return '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}';
    }
  }

  static String dateKeyFromDate(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

  static String current() => fromDate(DateTime.now());
}

File: web/lib/utils/month-key.ts
/** transactionDateKey = YYYY-MM-DD, monthKey = YYYY-MM — business dates, Asia/Karachi. */
export function monthKeyFromDate(d: Date): string {
  if (d.getDate() >= 10) {
    const next = new Date(d.getFullYear(), d.getMonth() + 1, 1);
    const y = next.getFullYear().toString().padStart(4, "0");
    const m = (next.getMonth() + 1).toString().padStart(2, "0");
    return `${y}-${m}`;
  } else {
    const y = d.getFullYear().toString().padStart(4, "0");
    const m = (d.getMonth() + 1).toString().padStart(2, "0");
    return `${y}-${m}`;
  }
}

export function dateKeyFromDate(d: Date): string {
  const y = d.getFullYear().toString().padStart(4, "0");
  const m = (d.getMonth() + 1).toString().padStart(2, "0");
  const day = d.getDate().toString().padStart(2, "0");
  return `${y}-${m}-${day}`;
}

export function currentMonthKey(): string {
  return monthKeyFromDate(new Date());
}
